weave/compile.py

```diff
@@ -496,36 +497,106 @@ def _needs_gql_propagation(node: graph.OutputNode) -> bool:
+        plugin = gql_op_plugin.get_gql_plugin(opdef.derived_from)
         unwrapped_first_arg_type = typing.cast(
             types.List, unwrapped_first_arg_type
         ).object_type
 
-    return fq_opname.endswith("GQLResolver") or (
+    return opdef.is_gql_root_resolver() or (
         plugin is not None
-        and plugin.gql_key_prop_fn is not None
-        and isinstance(unwrapped_first_arg_type, gql_with_keys.GQLHasKeysType)
+        and plugin.gql_op_output_type is not None
+        and isinstance(unwrapped_first_arg_type, partial_object.PartialObjectType)
     )
 
 
+def _initialize_gql_types_map_fn(node: graph.Node) -> typing.Optional[graph.Node]:
+    if isinstance(node, graph.OutputNode):
+        from_op = node.from_op
+
+        if from_op.name == "gqlroot-wbgqlquery":
+            # get the initial type
+            assert "query_str" in from_op.inputs and isinstance(
+                from_op.inputs["query_str"], graph.ConstNode
+            )
+
+            output_type = gql_to_weave.get_query_weave_type(
+                compile_domain.normalize_gql_query_string(
+                    from_op.inputs["query_str"].val
+                )
+            )
+
+            return graph.OutputNode(
+                output_type,
+                "gqlroot-wbgqlquery",
+                {
+                    **from_op.inputs,
+                    "output_type": graph.ConstNode(types.TypeType(), output_type),
+                },
+            )
+
+        if from_op.name == "gqlroot-querytoobj":
+            assert "gql_query_fragment" in from_op.inputs and isinstance(
+                from_op.inputs["gql_query_fragment"], graph.ConstNode
+            )
+            inner_fragment = from_op.inputs["gql_query_fragment"].val
+
+            assert "output_type" in from_op.inputs and isinstance(
+                from_op.inputs["output_type"], graph.ConstNode
+            )
+            output_type = from_op.inputs["output_type"].val
+
+            if isinstance(output_type, partial_object.PartialObjectTypeGeneratorType):
+                key_type = typing.cast(
+                    types.TypedDict,
+                    gql_to_weave.get_query_weave_type(
+                        compile_domain.normalize_gql_query_string(
+                            compile_domain.fragment_to_query(inner_fragment)
+                        )
+                    ),
+                )
+
+                key = gql_to_weave.get_outermost_alias(inner_fragment)
+                subtype = typing.cast(types.TypedDict, key_type.property_types[key])
+                output_type = output_type.with_attrs(subtype.property_types)
+
+                return graph.OutputNode(
+                    output_type,
+                    "gqlroot-querytoobj",
+                    {
+                        **from_op.inputs,
+                        "output_type": graph.ConstNode(types.TypeType(), output_type),
+                    },
+                )
+
+    return node
```

Comment:
Moved out of compile_gql_query to enforce separation of concerns. compile_gql is more or less unchanged now